lib/puppet/application/ssl.rb

```diff
@@ -80,8 +86,15 @@ def main
       raise Puppet::Error, _("An action must be specified.")
     end
 
-    Puppet.settings.use(:main, :agent)
```

Comment:
We still need to do this in the non-device case. I would expect puppet device needs something similar, including the `:device` section?

Reply:
I don't think we ever actually implemented the `:device` section. I've never seen one in the wild, and ...

https://puppet.com/docs/puppet/5.5/config_file_main.html#config-sections
https://github.com/puppetlabs/puppet/blob/master/lib/puppet/settings.rb#L996

Here is where that appears to have been introduced:

https://github.com/puppetlabs/puppet/commit/07b677c5f6af8def03c5c30393fd83bc3986239a#diff-04320a89c8d93fd690987c658c04d842R234

Reply:
> I would expect puppet device needs something similar, including the :device section?

Empirically, no:

I've tested `puppet ssl clean --target device.example.com` with and without either `Puppet.settings.use(:main, :agent)` or  `Puppet.settings.use(:main, :agent, :device)` after `host = Puppet::SSL::Host.new` and all succeed without error.

These are the only references to `:device` that I can find ...

https://github.com/puppetlabs/puppet/blob/master/lib/puppet/application/device.rb#L377
https://github.com/puppetlabs/puppet/blob/master/lib/puppet/defaults.rb#L1338

device.rb calls device in only one out three cases. Possibly @tphoney has insight into that.

Otherwise, this appears safest ...

```
    if options[:target]
      # Override the following, as per lib/puppet/application/device.rb
      Puppet[:certname] = options[:target]
      Puppet[:confdir]  = File.join(Puppet[:devicedir], Puppet[:certname])
      Puppet[:vardir]   = File.join(Puppet[:devicedir], Puppet[:certname])
      host = Puppet::SSL::Host.new(Puppet[:certname], true)
      Puppet.settings.use(:main, :agent, :device)
    else
      host = Puppet::SSL::Host.new
      Puppet.settings.use(:main, :agent)
    end
```

I'll push that.